include/boost/thread/concurrent_queues/detail/sync_queue_base.hpp

```diff
@@ -91,16 +91,14 @@ namespace detail
     inline bool wait_until_not_empty_or_closed(unique_lock<mutex>& lk);
     template <class WClock, class Duration>
     queue_op_status wait_until_not_empty_or_closed_until(unique_lock<mutex>& lk, chrono::time_point<WClock,Duration> const&tp);
-    template <class WClock, class Duration>
-    queue_op_status wait_until_closed_until(unique_lock<mutex>& lk, chrono::time_point<WClock,Duration> const&tp);
 
-    inline void notify_not_empty_if_needed(unique_lock<mutex>& )
+    inline void notify_elem_added(unique_lock<mutex>& )
     {
-      not_empty_.notify_one();
+      cond_.notify_all();
     }
-    inline void notify_not_empty_if_needed(lock_guard<mutex>& )
+    inline void notify_elem_added(lock_guard<mutex>& )
     {
-      not_empty_.notify_one();
+      cond_.notify_all();
```

Comment:
Because there may be more than one thread calling wait_to_pull*() and waiting for the next element to be ready. When a new element is added, both threads need to be notified so they can go around the for (;;) loop again and update how long they're waiting for.

Reply:
But seems only sync_timed_queue need to wake up all waiting thread to update new wait_until time point. If we always notify all when push, would it  degrade performance of  sync_queue or others?

Reply:
There may be a small degradation in performance with notify_all() if there are multiple threads waiting, but if there is only one thread waiting the performance of notify_one() and notify_all() should be about the same. And I still think notify_all() is necessary when there are multiple reader threads waiting to pull an element off of the sync_timed_queue.

Here is what may happen if we use notify_one() instead of notify_all(). Say you have a single writer thread (W) and two reader threads (R1 and R2).

- Time T1: W adds an item to the queue that's set to be pulled off at time T20.
- Time T2: R1 starts waiting to pull an item off of the queue. It calls cond_.wait_until(T20) and sleeps.
- Time T3: R2 starts waiting to pull an item off of the queue. It calls cond_.wait_until(T20) and sleeps.
- Time T4: W adds an item to the queue that's set to be pulled off at time T10. It wakes up **one** thread, and let's assume that it is R1. So R1 goes around the for(;;) loop and calls cond_.wait_until(T10) and sleeps.
- _Because R2 was never woken up, it is still waiting until time T20 and so has no chance of pulling the first element off of the queue. This is the situation I am trying to avoid. I think that both R1 and R2 should be waiting until time T10 and so both threads will have an equal chance of pulling the first item off of the queue. Does that make sense?_

Perhaps making sure that both R1 and R2 have an equal chance of pulling the first item off of the queue isn't important. I'm open to feedback if you and Vicente think it is not important.

Reply:
OK, understood, thanks a lot for explaining and quick fix of issue.